lib/domain.js

```diff
@@ -387,3 +382,49 @@ Domain.prototype.bind = function(cb) {
 
   return runBound;
 };
+
+// Override EventEmitter methods to make it domain-aware.
+EventEmitter.usingDomains = true;
+
+const eventInit = EventEmitter.init;
+EventEmitter.init = function() {
+  this.domain = null;
+  if (exports.active && !(this instanceof exports.Domain)) {
+    this.domain = exports.active;
+  }
+
+  return eventInit.call(this);
+};
+
+const eventEmit = EventEmitter.prototype.emit;
+EventEmitter.prototype.emit = function emit(...args) {
+  const domain = this.domain;
+  if (domain === null || domain === undefined || this === process) {
+    return eventEmit.apply(this, args);
```

Comment:
Could you use `Reflect.apply` here (and everywhere else) to avoid any potentially overridden `apply` methods?

Reply:
Why would we use `Reflect.apply` here but not in the regular event emitter code?

Reply:
There's a PR introducing it there too, it's just currently blocked on some other stuff.

Reply:
@apapirovski sounds good - although a little defensive (if someone overrides `Function.prototype.apply` or apply on a function object - they other didn't mean to or they know exactly what they're doing - in both cases I'm not sure Node.js should guard against it.

Reply:
We definitely should. The usage of `.apply` is an implementation detail and should not be affected by monkey-patching.

Reply:
Doesn't `Reflect.apply` have the same "issue"? It is writable too.

Reply:
Of course. But it's a step better because we're not longer relying on a method of a user provided function. At least `Reflect.apply` allows us consistent behaviour between all the different functions that are going to be called, instead of relying on the potentially unique apply of each specific callback.

Reply:
Also, to elaborate, this has become less of an issue lately but it was potentially even more problematic in the past where certain functions were called as `callback(...arguments)` and others `callback.apply(undefined, arguments)`.